Unity consumes messagepack as a nuget package instead of source

neuecc opened the pull request:
@AArnott  @pCYSl5EDgo 

This is a request for a significant change in the project structure. 
Currently, all the source code placed on the Unity side is to be moved to the .NET side, with Unity referencing a dll. 
This change will make it easier to write as a .NET library, and also allows us to adopt C# 12. 
There is no need to be conscious of Unity in regular code writing.

On the Unity side, it will reference a NuGet library built as a DLL as its core, 
and use a plugin method to separately reference extensions for Unity through Unity UPM.

MessagePack for C# is widely referenced as a foundational library, but 
since there is no compatibility between the Unity version and the NuGet version of the library, 
NuGet libraries that depend on MessagePack for C# in Unity have become unreferenceable. 
Unity is now different from the past, being .NET Standard 2.1, 
and excellent package managers like NuGetForUnity have also emerged.

I recently released a new library called R3, and there, I tried the same structure as this PR. 
https://github.com/Cysharp/R3/#unity

I believe adopting this PR is important for the efficiency of future development and the progress in Unity. 

However, since this would break compatibility with references on the Unity side, 
how about timing the release to coincide with the Source Generatorization when mpc is no longer needed?

AArnott commented:
That all sounds awesome. I'll review today or this weekend. I expect it'll cause massive merge conflicts with my work on #1691, so I'll suspend that work for now to not compound the problem.

AArnott commented:
Should we bump the major version number given that it'll be a breaking change for unity folks, as well as users of the analyzer?

AArnott commented:
How does this impact any hesitancy to take on other nuget dependencies? In the past, I think dependencies on other assemblies were taken only with huge justification because it complicated consumption for unity users.

pCYSl5EDgo commented:
> Should we bump the major version number given that it'll be a breaking change for unity folks, as well as users of the analyzer?

Major version 3 is appropriate because the `mpc` deletion is big breaking change.

neuecc commented:
> In the past, I think dependencies on other assemblies were taken only with huge justification because it complicated consumption for unity users.

Disable Assembly Version Validation solves it.
![image](https://github.com/MessagePack-CSharp/MessagePack-CSharp/assets/46207/d589351c-ca93-4e04-bc7d-36c4fc380552)

II also think 3 is a good choice. Converting to Source Generator would be a significant change for all users.

---

This draft is rough and there are several tasks remaining.

* Need to remove code like `if UNITY` or `IF IL2CPP` from the Core
* Especially for Unity Android, there was code tailored for binary Write/Read, and we need to decide what to do with that
* There was an issue with sharing test code because signed `InternalVisibleTo` couldn't be referenced on the Unity side. (It was only for `Sequence`, so I temporarily pasted the `Sequence` itself to remove the compile error.)

---

Since NuGetForUnity also automatically reflects Source Generators, creating SourceGenerators for both 3.x and 4.x and including the dlls should automatically segregate them, which would be good. Let's not create a SourceGenerator package specifically for Unity.

By the way, Unity supports Incremental Generators depending on the version, and currently, it seems that versions 4.1.0(Unity 2022.2) and 4.3.0(latest) are available.

AArnott commented:
All this sounds exciting. Do you mind holding off on this substantial refactoring until my source generator/analyzer work to switch to attributes is done?

pCYSl5EDgo commented:
Dropping `netstandard2.0` from `MessagePack` csprojs (except Source Generator) will be permitted?

AArnott commented:
No, we can't drop netstandard2.0. I believe what neuecc is saying is because unity supports netstandard2.1, we don't need a special source-based build for it any more.

neuecc commented:
> Do you mind holding off on this substantial refactoring until my source generator/analyzer work to switch to attributes is done?

OK.

pCYSl5EDgo commented:
> Especially for Unity Android, there was code tailored for binary Write/Read, and we need to decide what to do with that

I've installed Unity2022.3.18 and built IL2CPP in order to assess the cost of runtime Android armv7 detection.
All of the latter C++ codes are generated with windows x64 master build settings.

### Results

- `BitConverter.IsLittleEndian` is compiled to be an inlined function returning constant boolean.
- `sizeof(nint) == 4` is compiled to a constant-like expression.
- `RuntimeInformation.ProcessArchitecture == Architecture.Arm` is compiled to the code which accesses static field.

Considering the compile-time optimization of C++, the combination of `BitConverter.IsLittleEndian` and `sizeof(nint) == 4` is the best method. Even if this method cannot distinguish x86-windows and armv7-android, it has the best runtime efficiency.

### Methods

<details><summary>Test C# code</summary>

```csharp
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using UnityEngine;

public unsafe class TestForConst : MonoBehaviour
{
    void Update()
    {
        Vector3 velocity;
        if (BitConverter.IsLittleEndian)
        {
            velocity = new(0, 1, 0);
        }
        else if (sizeof(nint) == 4)
        {
            velocity = new(0, -1, 0);
        }
        else
        {
            velocity = new();
        }

        velocity += GenericVector<int>.Get();
        
        transform.position += velocity * Time.deltaTime;
    }
}

internal struct GenericVector<T>
{
    public static Vector3 Get()
    {
        if (typeof(T) == typeof(int))
        {
            if (RuntimeInformation.ProcessArchitecture == Architecture.Arm)
            {
                return new(1, 0, 1);
            }

            return new(1, 0, 0);
        }

        return new(-1, 0, 0);
    }
}
```

</details>

<details><summary>Generated C++ code for Update method</summary>

```c++
IL2CPP_EXTERN_C IL2CPP_METHOD_ATTR void TestForConst_Update_mAECC56A194C2E9D162EAFFB3B9F23B0B4EBCC9F7 (TestForConst_tDA600D04A981E1B53CD873D2147332DB5AA8480E* __this, const RuntimeMethod* method) 
{
	static bool s_Il2CppMethodInitialized;
	if (!s_Il2CppMethodInitialized)
	{
		il2cpp_codegen_initialize_runtime_metadata((uintptr_t*)&GenericVector_1_Get_m49BFA61A2FA9A6589F796D73B626ED5C0562B4A7_RuntimeMethod_var);
		s_Il2CppMethodInitialized = true;
	}
	Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 V_0;
	memset((&V_0), 0, sizeof(V_0));
	{
		if (!il2cpp_codegen_is_little_endian())
		{
			goto IL_001f;
		}
	}
	{
		Vector3__ctor_m376936E6B999EF1ECBE57D990A386303E2283DE0_inline((&V_0), (0.0f), (1.0f), (0.0f), NULL);
		goto IL_0048;
	}

IL_001f:
	{
		uint32_t L_0 = sizeof(intptr_t);
		if ((!(((uint32_t)L_0) == ((uint32_t)4))))
		{
			goto IL_0040;
		}
	}
	{
		Vector3__ctor_m376936E6B999EF1ECBE57D990A386303E2283DE0_inline((&V_0), (0.0f), (-1.0f), (0.0f), NULL);
		goto IL_0048;
	}

IL_0040:
	{
		il2cpp_codegen_initobj((&V_0), sizeof(Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2));
	}

IL_0048:
	{
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_1 = V_0;
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_2;
		L_2 = GenericVector_1_Get_m49BFA61A2FA9A6589F796D73B626ED5C0562B4A7(GenericVector_1_Get_m49BFA61A2FA9A6589F796D73B626ED5C0562B4A7_RuntimeMethod_var);
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_3;
		L_3 = Vector3_op_Addition_m78C0EC70CB66E8DCAC225743D82B268DAEE92067_inline(L_1, L_2, NULL);
		V_0 = L_3;
		Transform_tB27202C6F4E36D225EE28A13E4D662BF99785DB1* L_4;
		L_4 = Component_get_transform_m2919A1D81931E6932C7F06D4C2F0AB8DDA9A5371(__this, NULL);
		Transform_tB27202C6F4E36D225EE28A13E4D662BF99785DB1* L_5 = L_4;
		NullCheck(L_5);
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_6;
		L_6 = Transform_get_position_m69CD5FA214FDAE7BB701552943674846C220FDE1(L_5, NULL);
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_7 = V_0;
		float L_8;
		L_8 = Time_get_deltaTime_mC3195000401F0FD167DD2F948FD2BC58330D0865(NULL);
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_9;
		L_9 = Vector3_op_Multiply_m87BA7C578F96C8E49BB07088DAAC4649F83B0353_inline(L_7, L_8, NULL);
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_10;
		L_10 = Vector3_op_Addition_m78C0EC70CB66E8DCAC225743D82B268DAEE92067_inline(L_6, L_9, NULL);
		NullCheck(L_5);
		Transform_set_position_mA1A817124BB41B685043DED2A9BA48CDF37C4156(L_5, L_10, NULL);
		return;
	}
}
```

</details>

<details><summary>il2cpp_codegen_is_little_endian is well optimized.</summary>

```c++
inline bool il2cpp_codegen_is_little_endian()
{
#if IL2CPP_BYTE_ORDER == IL2CPP_LITTLE_ENDIAN
    return true;
#else
    return false;
#endif
}
```

</details>

<details><summary>`typeof(T) == typeof(int)` is bad.</summary>

```c++
IL2CPP_EXTERN_C IL2CPP_METHOD_ATTR Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 GenericVector_1_Get_m49BFA61A2FA9A6589F796D73B626ED5C0562B4A7_gshared (const RuntimeMethod* method) 
{
	static bool s_Il2CppMethodInitialized;
	if (!s_Il2CppMethodInitialized)
	{
		il2cpp_codegen_initialize_runtime_metadata((uintptr_t*)&Int32_t680FF22E76F6EFAD4375103CBBFFA0421349384C_0_0_0_var);
		il2cpp_codegen_initialize_runtime_metadata((uintptr_t*)&RuntimeInformation_tB2DFA85FB9251AE3A3112904C9DF06C30D1D3EAF_il2cpp_TypeInfo_var);
		il2cpp_codegen_initialize_runtime_metadata((uintptr_t*)&Type_t_il2cpp_TypeInfo_var);
		s_Il2CppMethodInitialized = true;
	}
	{
		RuntimeTypeHandle_t332A452B8B6179E4469B69525D0FE82A88030F7B L_0 = { reinterpret_cast<intptr_t> (il2cpp_rgctx_type(InitializedTypeInfo(method->klass)->rgctx_data, 0)) };
		il2cpp_codegen_runtime_class_init_inline(Type_t_il2cpp_TypeInfo_var);
		Type_t* L_1;
		L_1 = Type_GetTypeFromHandle_m6062B81682F79A4D6DF2640692EE6D9987858C57(L_0, NULL);
		RuntimeTypeHandle_t332A452B8B6179E4469B69525D0FE82A88030F7B L_2 = { reinterpret_cast<intptr_t> (Int32_t680FF22E76F6EFAD4375103CBBFFA0421349384C_0_0_0_var) };
		Type_t* L_3;
		L_3 = Type_GetTypeFromHandle_m6062B81682F79A4D6DF2640692EE6D9987858C57(L_2, NULL);
		bool L_4;
		L_4 = Type_op_Equality_m99930A0E44E420A685FABA60E60BA1CC5FA0EBDC(L_1, L_3, NULL);
		if (!L_4)
		{
			goto IL_004d;
		}
	}
	{
		il2cpp_codegen_runtime_class_init_inline(RuntimeInformation_tB2DFA85FB9251AE3A3112904C9DF06C30D1D3EAF_il2cpp_TypeInfo_var);
		int32_t L_5;
		L_5 = RuntimeInformation_get_ProcessArchitecture_mB2DAF77FAF4F8F97AE4045FA7DD140D60D8BF3F3_inline(NULL);
		if ((!(((uint32_t)L_5) == ((uint32_t)2))))
		{
			goto IL_0038;
		}
	}
	{
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_6;
		memset((&L_6), 0, sizeof(L_6));
		Vector3__ctor_m376936E6B999EF1ECBE57D990A386303E2283DE0_inline((&L_6), (1.0f), (0.0f), (1.0f), NULL);
		return L_6;
	}

IL_0038:
	{
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_7;
		memset((&L_7), 0, sizeof(L_7));
		Vector3__ctor_m376936E6B999EF1ECBE57D990A386303E2283DE0_inline((&L_7), (1.0f), (0.0f), (0.0f), NULL);
		return L_7;
	}

IL_004d:
	{
		Vector3_t24C512C7B96BBABAD472002D0BA2BDA40A5A80B2 L_8;
		memset((&L_8), 0, sizeof(L_8));
		Vector3__ctor_m376936E6B999EF1ECBE57D990A386303E2283DE0_inline((&L_8), (-1.0f), (0.0f), (0.0f), NULL);
		return L_8;
	}
}
```

</details>

pCYSl5EDgo commented:
I've tested `Unsafe.ReadUnaligned` and found that Unity2022.3 treats Unsafe.ReadUnaligned as special.
It seems that we no longer need to take care of Android armv7 problem?

```c++
template<typename T>
inline T il2cpp_unsafe_read_unaligned(void* location)
{
    T result;
#if IL2CPP_TARGET_ARMV7 || IL2CPP_TARGET_JAVASCRIPT
    memcpy(&result, location, sizeof(T));
#else
    result = *((T*)location);
#endif
    return result;
}

template<typename T>
inline void il2cpp_unsafe_write_unaligned(void* location, T value)
{
#if IL2CPP_TARGET_ARMV7 || IL2CPP_TARGET_JAVASCRIPT
    memcpy(location, &value, sizeof(T));
#else
    *((T*)location) = value;
#endif
}

template<typename T, typename TOffset>
inline T* il2cpp_unsafe_add(void* source, TOffset offset)
{
    return reinterpret_cast<T*>(source) + offset;
}

template<typename T, typename TOffset>
inline T* il2cpp_unsafe_add_byte_offset(void* source, TOffset offset)
{
    return reinterpret_cast<T*>(reinterpret_cast<uint8_t*>(source) + offset);
}
```

Unfortunately, Unity does not generate intrinsics for `MemoryMarshal.GetReference` and `Unsafe.AddByteOffset`.
There seems to be `il2cpp_unsafe_add_byte_offset` but not used even if `Unsafe.AddByteOffset` is used!
In contrast, `il2cpp_unsafe_add` is used when `Unsafe.Add` is used.
Very surprising.

AArnott commented:
I think it's time to focus on this PR. Since I introduced all the merge conflicts, I'll try to find time to freshen it up.

AArnott commented:
Ok, it's freshened up, ready for someone with more Unity experience to patch up the rest. :)

AArnott commented:
@neuecc and @pCYSl5EDgo please review #1825 as an early draft for getting rid of the unity-related `#if` regions.

neuecc commented:
sorry for awaying check PRs.
still active and needs any check?

AArnott commented:
Good to have you back, @neuecc. Yes, any post mortem comments you have on #1825 are welcome. As well as further guidance on how to remove the remaining unity bits from the codebase in this PR.

neuecc commented:
Wow, it's great to see that the mono/IL2CPP bug and workaround hacks are no longer needed.
I'm happy to think that this project may have contributed to that.

It would be unfortunate if this refactoring becomes a blocker for v3.
I will also take a look into it.

neuecc commented:
First, I intend to normalize the management of MessagePack.UnityClient. 
The new version of NuGetForUnity includes an important feature (prioritizing netstandard2.1), so I updated it.
Also, by running `dotnet nugetforunity restore` in local tools, dependent assemblies can be obtained,
so it seems good to execute it in init.bat/build.bat instead.
Furthermore, this allows the management of dependent DLLs to be removed from Assets.
I'm also considering a few other adjustments.

The handling of the dependency on Microsoft.Bcl.AsyncInterfaces varies depending on whether the target is set to netstandard2.0 or netstandard2.1.
I think it's better to reference netstandard2.1.

AArnott commented:
> The handling of the dependency on Microsoft.Bcl.AsyncInterfaces varies depending on whether the target is set to netstandard2.0 or netstandard2.1.
I think it's better to reference netstandard2.1.

Reference it from where? Unity always uses netstandard2.1 at this point, right? 

I have no objection to targeting netstandard2.1, but we should continue to multi-target netstandard2.0 and .NET itself as well.

neuecc commented:
my first setup(and current) reference 2.0, I've changed to 2.1.
https://github.com/MessagePack-CSharp/MessagePack-CSharp/pull/1734/commits/14e63e7ef08ac2abf8f6ad070930ff05bba2e203

neuecc commented:
* revert unity specified code from UnityShims to UnityClient, UnityShims is shims, can not use in Unity so needs concrete code in Unity.
* changed internal ThisLibraryExtensionTypeCodes -> public ReservedExtensionTypeCodes because UnityClient reference this type.
  * By the way, what's the best way to add content to Unshiped.txt? There are five of them, which seems quite daunting, so I haven't been adding much.
* Update Unity Editor version to latest 2021.3(2021.3.39f1), 2021.3.11f1 has bugs of IL2CPP build on Windows( https://forum.unity.com/threads/workaround-for-building-with-il2cpp-with-visual-studio-2022-17-4.1355570/ )

Tomorrow, I will investigate how Source Generators work with IL2CPP and consider ways to add default resolvers in Unity. Also, the remaining #if UNITY_2018_3_OR_NEWER directives on the .NET side are confusing, so We should remove them all.

AArnott commented:
> what's the best way to add content to Unshiped.txt? There are five of them, which seems quite daunting, so I haven't been adding much.

Within Visual Studio, navigate to one of the warning lines (double-clicking on the Error List is easy) then press Ctrl+. (or otherwise activate the popup button) and click the command to apply the fix. If you select to apply the bulk fix for the whole Solution, all 5 issue will be fixed at once.

neuecc commented:
Thank you, I made the correction of unshipped.
However, for net8.0, it seems the correction only applies to net8.0, so I needed to make corrections for each targetframework: net472, net6.0, netstandard2.0, and netstandard2.1.

Status of SourceGenerator in Unity

* 2021.3 doesn't work with Incremental Generator at all, so I switched to 2022.3.12f1
* Source Generator doesn't work in Unity, but v3.0.54-alpha does work
   * After investigation, the cause was that the reference to Immutable Collections was updated to 7.0.0
   * So I removed these, and to avoid potential dependencies in Unity, I removed all of them as a precaution
     * C# polyfill, using PolySharp instead
     * https://github.com/MessagePack-CSharp/MessagePack-CSharp/pull/1734/commits/c5ba1ca3fac407021b922804d808fe4972a675c6#diff-5baf5f9e448ad54ab25a091adee0da05d4d228481c9200518fcb1b53a65d4156
     * https://github.com/MessagePack-CSharp/MessagePack-CSharp/pull/1734/commits/c5ba1ca3fac407021b922804d808fe4972a675c6#diff-f2b278f0b2f3068197b2ba61250b32ce2cab7a9ee0b93d95d2629e0b0ec8c7f2
* In the current commit, it works in Unity, but now MessagePack.Analyzers can't build
   * It's detecting a downgrade because System.Collections.Immutable hasn't been updated to 7.0.0
   * Is it because MessagePack.Analyzers references Microsoft.CodeAnalysis.Common 4.8.0?
   * Isn't MessagePack.Analyzers essentially an empty project? Also, since it's IsAnalyzerProject, shouldn't the reference be 4.3.0 even if there is one?
   * I'm not sure about this part, so could you help with this correction? @AArnott

AArnott commented:
> However, for net8.0, it seems the correction only applies to net8.0, so I needed to make corrections for each targetframework: net472, net6.0, netstandard2.0, and netstandard2.1.

Ya, if you bulk fix the Solution it's _supposed_ to apply the fix to all the required Unshipped files, but lately I've seen it fail to. In order to fix the other target frameworks, you have to switch the active target framework in the editor to each one and apply the fix to each.

AArnott commented:
> In the current commit, it works in Unity, but now MessagePack.Analyzers can't build

I'll take a look today.

> I removed these, and to avoid potential dependencies in Unity, I removed all of them as a precaution
C# polyfill, using PolySharp instead

That's potentially a problem, because PolySharp has a serious bug, which I just filed:
https://github.com/Sergio0694/PolySharp/issues/98
It is imperative that we not use PolySharp on any assembly used as a reference assembly to another. For source generator projects, it's probably not going to be a problem. But for other projects in the repo, it could be. I suggest we remove all use of PolySharp and switch back to the individual polyfill packages that we were using before, which don't have this bug. They don't represent runtime dependencies any more than PolySharp does, so I don't expect it will present a problem for Unity.

Sergio0694 commented:
> "It is imperative that we not use PolySharp on any assembly used as a reference assembly to another. For source generator projects, it's probably not going to be a problem. But for other projects in the repo, it could be. I suggest we remove all use of PolySharp and switch back to the individual polyfill packages that we were using before, which don't have this bug. They don't represent runtime dependencies any more than PolySharp does, so I don't expect it will present a problem for Unity."

I commented in the other issue as well, I don't think this is actually an issue. PolySharp generates all needed type forwards.

AArnott commented:
@neuecc I've pushed a fix for the source generator package downgrade warning. I expect it will still work in unity but haven't tested it.

We have other build breaks because in c5b2e8c8c127f4390459602c93d2cbbc27c3c928 you commented out the `Class1.cs` file, which the tests use.

neuecc commented:
Thank you!

* To use Unity practically, UnityResolver needs to be inserted into the Resolver by default
  * Decided to use Unity's RuntimeInitializeOnLoadMethod to set `MessagePackSerializer.DefaultOptions = MessagePackSerializerOptions.Standard.WithResolver(UnityResolver.InstanceWithStandardResolver)` at startup
  * `SourceGeneratedFormatterResolver` seems to work fine with IL2CPP, so regular users don't need to touch the Options at all
* Sharing UnitTests with the .NET version seemed difficult, so I removed all of them this time
  * All xUnit features can be used on the .NET side
  * As we're using the same DLL, basic functionality should be guaranteed, so there shouldn't be any major issues
  * Of course, having tests would be better for Unity-specific cases like IL2CPP
  * For now, I've added only basic functionality verification tests
* I want to run the above IL2CPP functionality verification tests in CI
  * We've migrated to GitHub Actions, but Unity 2022 license authentication is needed
  * I'll work on this tomorrow, and with that will finish this branch
* All `#if UNITY_` code has been removed
  * https://github.com/MessagePack-CSharp/MessagePack-CSharp/pull/1734/commits/b40d7e7767670de9b784681ae0429d202a7b2e15

By the way, why is the `CodeFixes.csproj` fix not valid if it is in the csproj, but only if it is in the `Directory.Build.props`?
I thought I thought ArtifactsPath and others had a similar spec, but I don't understand this, so I'd appreciate it if you could enlighten me.

AArnott commented:
> Decided to use Unity's RuntimeInitializeOnLoadMethod to set

That sounds great. But if unity developers will be consuming the package, how/when will they get the source code that you're adding into the scripts directory?

> By the way, why is the CodeFixes.csproj fix not valid if it is in the csproj, but only if it is in the Directory.Build.props?

It's an unfortunate consequence of msbuild evaluation order. The `IsAnalyzerProject` property is read from `Directory.Packages.props`, which is imported before the content of CodeFixes.csproj itself is evaluated. But `Directory.Build.props` is evaluated "on the way down" to Directory.Packages.props, so setting the property there works.

Another option would be to restructure the `CodeFixes.csproj` like this:

```xml
<Project>
   <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />
   <!-- rest of the current project content -->
   <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.targets" />
</Project>
```

And then we could move the `IsAnalyzerProject` property definition _above_ the first Sdk import.

neuecc commented:
> But if unity developers will be consuming the package, how/when will they get the source code that you're adding into the scripts directory?

Either way, Unity users will need to import UnityResolver as source code, so they will install two packages:
1. Import `MessagePack` using NuGetForUnity (I believe this includes the Source Generator, is that correct?)
2. Import `MessagePack.Unity` using UPM (git reference), which includes UnityResolver and changes to default options via RuntimeInitializeOnLoadMethod

> `Directory.Build.props` is evaluated "on the way down" to Directory.Packages.props, so setting the property there works.

Thank you, I understand!

AArnott commented:
> Import MessagePack using NuGetForUnity (I believe this includes the Source Generator, is that correct?)

Yes, assuming `NuGetForUnity` brings down dependencies too, and assuming it knows how to recognize roslyn analyzers and register them with unity, then the source generators will Just Work.

neuecc commented:
@AArnott 

I have completed all the changes, and I think it's now safe to merge.

* Set up Unity IL2CPP build on Linux CI and run the built unit tests
  * This is a good step forward as we didn't have IL2CPP CI running before!
* Modified the Unity section in the ReadMe
  * https://github.com/MessagePack-CSharp/MessagePack-CSharp/pull/1734/commits/9845d3f44734059ef3522f761bea64c74cd7c420
  * There are also some changes related to the Source Generator text, so please revise if further modifications are needed

By the way, is it possible to generate Source Generator externally in a library that only references the Annotations package?
It seems like it might not be possible based on the current setup, but what do you think?

I'm not sure if this is possible, but how about something like the following specifications to allow specifying external types?

```csharp
// Search for [MessagePackObject] types from the assembly containing Foo
[GeneratedMessagePackResolver(Assembly: typeof(Foo))]
public partial class GeneratedMessagePackResolver
{
}

// Specify types to generate (Creating with Map might allow non-[MessagePackObject] types?)
[GeneratedMessagePackResolver(Types: [typeof(Foo), typeof(Bar)])]
public partial class GeneratedMessagePackResolver
{
}
```

AArnott commented:
> I have completed all the changes, and I think it's now safe to merge.

Super cool. I'll review today.

>  is it possible to generate Source Generator externally in a library that only references the Annotations package?
It seems like it might not be possible based on the current setup, but what do you think?

The source generator is already in a separate package. That package has no dependencies at all (not even annotations).
But it's meaningless for a user to reference just that package, because without annotations you can't use attributes and thus there is nothing to analyze or generate. 
And if you _do_ reference the annotations library as well, while the analyzer part would theoretically be worthwhile, the source generator is pointless because the generated source will need the full library since the code implements and references all sorts of APIs from the main library. 
Since the source generator is 'on by default' in v3 for all attributes it finds, this makes referencing the analyzer/source generator package alone completely pointless. 

For that matter, since MessagePack depends on MessagePackAnalyzer in order for the 'on by default' behavior, and MessagePackAnalyzer by itself is pointless, that actually begs the question as to why we don't just merge the content of MessagePackAnalyzer into the MessagePack package and call it one. That's outside the scope of _this_ PR, but it's something we should settle one way or the other before v3 goes stable.

> // Search for [MessagePackObject] types from the assembly containing Foo
> `[GeneratedMessagePackResolver(Assembly: typeof(Foo))]`

I guess the idea there is to generate AOT formatters for data types in an assembly that wasn't compiled against v3, right? Sounds useful. If at runtime the other assembly _does_ provide its own formatters (because at runtime a newer dll is provided) I wonder which formatters should be used. I can imagine arguments for either way. Do you want to open an issue in the v3 milestone to track the discussion?

> // Specify types to generate (Creating with Map might allow non-[MessagePackObject] types?)
> [GeneratedMessagePackResolver(Types: [typeof(Foo), typeof(Bar)])]

I really don't want to expand on support for contractless scenarios (which is what it sounds like you're talking about) until we have a really thorough design for it. Contractless today has a poor success rate in my experience (and based on the issues that keep coming in), and if we're going to have AllowPrivate behavior by default (which I'm very excited about), ContractlessAllowPrivate is even worse, perhaps, unless we think through it carefully. Let's consider that in a new issue. And it sounds like something we could add in v3.1 as it wouldn't be a breaking change.

AArnott reviewed (commented):
I see you removed `prepare_release.ps1`. If that means we no longer have a version number in some file to keep in sync with `version.json` for the sake of the unity package, that's great.

AArnott reviewed (changes requested):
Looks very nearly ready. Very exciting!

neuecc commented:
Oh, we do need `prepare_release.ps1`. I deleted too much.
We still need to rewrite the version here to synchronize it with the release!
https://github.com/MessagePack-CSharp/MessagePack-CSharp/blob/new-structure/src/MessagePack.UnityClient/Assets/Scripts/MessagePack/package.json#L4

neuecc commented on bin/MessagePack.SourceGenerator/Debug/netstandard2.0/package.json:
```diff
@@ -0,0 +1,17 @@
+        "name": "MessagePack-CSharp",
+        "url": "https://github.com/MessagePack-CSharp/MessagePack-CSharp"
+    },
+    "version": "1.0.0",
```
This is necessary for referencing pre-built DLLs from the Unity side for development purposes.
`MessagePack/Debug/netstandard2.1` is also processed similarly.
The version has no significance; it's purely for reference during development and is not used at all during release.
We're using Debug instead of Release because typically, during normal development, Debug is where most changes occur.

AArnott replied:
I think that answer would be great to put into a README in this directory, since otherwise I would think this was a mistake and delete them from git.

neuecc replied:
I added a note to the ReadMe in UnityClient and a guide to it from CONTRIBUTING.md.

neuecc commented on src/MessagePack.UnityClient/Assets/NuGet.config:
```diff
@@ -0,0 +1,16 @@
+<configuration>
+  <packageSources>
+    <clear />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
```
Since this is generated by NuGetForUnity, it doesn't have much deep meaning.
This is also for development purposes, so it's not included when users reference it.

neuecc commented on src/MessagePack.UnityClient/Assets/packages.config:
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
```
Users can resolve dependencies through NuGetForUnity,
but this project itself pulls dlls from bin/Debug,
so because dependencies can't be resolved, we're manually inserting them.

While it would be best if we could synchronize with the csproj,
this is only for running during development (necessary when running unit tests),
so I think it's probably okay if we're not too strict about versions.

neuecc commented on src/MessagePack.UnityClient/README.md:
```diff
@@ -0,0 +1,11 @@
+
+> dotnet nugetforunity restore
+
+Also, since the MessagePack library for development is referenced from `bin\MessagePack\Debug\netstandard2.1`, you need to generate the DLL with a debug build beforehand.
```
Yes, we need to build the main DLL itself (otherwise we'd have to upload to NuGet every time during development!)
Would it have been clearer if I had written that as a procedure, you need to run `dotnet build -c DEBUG` in the sln folder?
(By the way, the GitHub Actions CI follows exactly this procedure)

For releases, users will:
* Reference the pre-built dll from NuGet for `MessagePack/MessagePackAnalyzer` using NuGetForUnity
* Reference `src/MessagePack.UnityClient/Assets/Scripts/MessagePack/` as source code in UPM

AArnott replied:
You can change the doc if you feel it necessary. Don't mind my ignorant questions too much as I'm just not familiar with unity.
But I thought one of the value props from this new repo structure was that unity users would _always_ consume MessagePack.dll from the nuget package rather than build it themselves. I thought that was why we would be able to use the latest C# lang version going forward. 
Since unity users don't have to rebuild from source for dependencies of messagepack.dll, why is it that they have to build messagepack.dll itself?

AArnott replied:
Or are you talking about running the messagepack tests within unity as part of our own local validation or test pipeline? If that's the case, and if that doesn't really require users to build messagepack.dll themselves, then I'd like to explore how we can build from source for our own testing while providing the necessary .unitypackage for outside this repo to consume the messagepack.dll as a binary and get a few source files necessary for unity environments.

neuecc replied:
Yes, Unity users don't need to build, as they obtain the built dll from the official NuGet.

However, the developers of this repository are different; the dll under development doesn't exist on NuGet.
I think it's natural to use what's been built locally.
In regular .NET, a separate csproj (MessagePack) associated with the csproj would
automatically build and generate a DEBUG build dll to be retrieved.
This is what we want to do in this Unity project, and since
there's no means to automatically resolve dependencies and build, we're building manually.
Although it's manual, during normal development, you'd typically run the .NET side before looking at the Unity side,
so Unity would almost automatically get the latest build of the dll under development.

By the way, .unitypackage is completely irrelevant.
In this new structure, no .unitypackage should appear at all.

neuecc commented:
I think it's good that MessagePackAnalyzer is being merged into MessagePack.

Until now, a common practice for projects defining container types was
to have fewer dependencies (for example, using DataContract or referencing only the separated Annotations).
I believe there were quite a few people who structured their projects this way.
It's possible that those who had separated out the Annotations might raise complaints or questions about Source Generator support.

AArnott commented:
> It's possible that those who had separated out the Annotations might raise complaints or questions about Source Generator support.

Those who want the attributes and nothing else can continue to get that... but they will not be able to benefit from all the AOT work -- just like they couldn't with v2 of the library, since fundamentally the AOT formatters need to reference the main library.
